main.go

```diff
@@ -0,0 +1,335 @@
+		return
+	}
+	next(w,r)
+}
+
+func(info User)loginContactHandler (w http.ResponseWriter , r *http.Request){
+	var p LoginPage
+
+	if r.FormValue("signUp") != ""{
+		secret , _ := bcrypt.GenerateFromPassword([]byte(r.FormValue("password")) , bcrypt.DefaultCost)
+		info = User{r.FormValue("username") , secret}
+		stmt ,err := db.Prepare("INSERT Users SET username=? , userpassword=?")
+		_ ,err =stmt.Exec(info.Username ,info.Password)
+		if err != nil{
+			p.Error = err.Error()
+
+		}else{
+			//sessions.getsession() to create session variable
+			//.set key-> User , value ->user.Username
+			sessions.GetSession(r).Set("User" , info.Username)
+			http.Redirect(w, r, "/Home" , http.StatusFound)
+			return
+		}
+
+	}else if r.FormValue("login") != ""{
+		row ,err := db.Query("SELECT * FROM Users WHERE username =?" , r.FormValue("username"))
+		for  row.Next(){
+			row.Scan(&info.Username , &info.Password)
+		}
+		defer row.Close()
+		if err != nil {
+			p.Error = err.Error()
+			return
+		}else if row == nil{
+			p.Error = " No such user found with the username : " + r.FormValue("username")
+		}else {
+			err := bcrypt.CompareHashAndPassword(info.Password , []byte(r.FormValue("password")))
+			if err != nil{
+				p.Error = err.Error()
+			}else {
+				sessions.GetSession(r).Set("User" , info.Username)
+				http.Redirect(w, r, "/" , http.StatusFound)
+				return
+			}
+		}
+	}
+
+
+
+	templates := template.Must(template.ParseFiles("Login.html"))
+	err:= templates.Execute(w,p)
+	if err != nil {
+		http.Error(w, err.Error() , http.StatusInternalServerError)
+		return
+	}
+
+
+
+ 	 //info.username = r.FormValue("username")
+	 //info.password = r.FormValue("password")
+	//http.Redirect(w, r, "/Home" , http.StatusFound)
+}
+func(u User)logoutHandler(w http.ResponseWriter , r *http.Request){
+	sessions.GetSession(r).Set("User" , nil)
+	http.Redirect(w , r , "/Login" , http.StatusFound)
+}
+func(contact contact_info)SaveContactHandler(w http.ResponseWriter , r *http.Request){
+	contact.Name = r.FormValue("name")
+	contact.Number = r.FormValue("number")
+	contact.Email = r.FormValue("email")
+	contact.Nationality = r.FormValue("nationality")
+	contact.Address = r.FormValue("address")
+	contact.Username = sessions.GetSession(r).Get("User").(string)
+
+	stmt , err := db.Prepare("INSERT Contacts SET name=?  , email=? , nationality=? ,address=? ,username=?")
+	checkErr(err)
+	res , err := stmt.Exec(contact.Name,contact.Email,contact.Nationality,contact.Address,contact.Username)
+	checkErr(err)
+	id , err := res.LastInsertId()
+	checkErr(err)
+
+	var num Telephone
+	num.Number = r.FormValue("number")
+
+	num.Contact_id = int(id)
+
+	_ =addNum(num.Number , num.Contact_id)
+	sessions.GetSession(r).Set("Contact" , id)
+}
+func(contact contact_info)DeleteContactHandler (w http.ResponseWriter , r *http.Request){
+	ID , _ := strconv.ParseInt(mux.Vars(r)["id"] , 10 , 64)
+	fmt.Println(ID)
+	stmt , err :=db.Prepare("delete from Contacts where id=?")
```

Comment:
At SQL script , i added "on delete cascade" , which automatically if i deleted a contact it will delete all his phone numbers .